ledger/participant-integration-api/src/main/scala/platform/store/backend/common/EventStorageBackendTemplate.scala

```diff
@@ -21,14 +20,15 @@ import com.daml.platform.store.Conversions.{
 }
 import com.daml.platform.store.SimpleSqlAsVectorOf.SimpleSqlAsVectorOf
 import com.daml.platform.store.appendonlydao.events.{EventsTable, Identifier, Raw}
-import com.daml.platform.store.backend.EventStorageBackend
+import com.daml.platform.store.backend.{EventStorageBackend, ParameterStorageBackend}
 import com.daml.platform.store.backend.EventStorageBackend.{FilterParams, RangeParams}
 import com.daml.platform.store.backend.StorageBackend.RawTransactionEvent
 import com.daml.platform.store.backend.common.ComposableQuery.{CompositeSql, SqlStringInterpolation}
 
 import scala.collection.compat.immutable.ArraySeq
 
 trait EventStorageBackendTemplate extends EventStorageBackend {
+  this: ParameterStorageBackend =>
```

Comment:
@rautenrieth-da @nmarton-da . Added this self type here since we need to query the `participant_all_divulged_contracts_pruned_up_to_inclusive` in the `pruneEvents` here. Not sold on it, but the alternative would be to execute this query before the call to `pruneEvents` and pass it as a parameter, even though it might not be needed. Wdyt?

Reply:
I would keep the self type here.

Reply:
@tudor-da can you please add a template method instead? that is also an ugly inheritance code sharing construct, but at least limits the surface of the dependency.
```
def participantAllDivulgedContractsPrunedUpToInclusive(
      connection: Connection
  ): Option[Offset]
```

The jury is still out regarding code sharing between storage backend's, but some thoughts on that matter:
* It would be nice if we wont have to
* It is on the table to switch to composition instead of inheritance here
* Generally the current direction seem to be that storage backend methods seem to be rather small and dumb, therefore code sharing is generally something we can avoid. In your example the `pruneEvents` is rather big and smart, and maybe that is the cause for this dependency hunger here.
* One possible direction (which we also followed with @rautenrieth-da) is to keep the StorageBackend functions small and dumb, and do composition one layer above (what usually is JdbcLedgerDao, but can be something else as well) -- in your case that would mean to rip apart the `pruneEvents` to smaller functions, and move the conditional logic above, which removes the need to share code on storage-backend layer (I absolutely won't do that in this PR, especially because it would involve move of test code as well)
* One other direction would be to change grouping semantics (create a PruningStorageBackend, and move stuff there from EventStorageBackend and ParameterStorageBackend...and maybe from other places), which removes the need again for code sharing between StorageBackend silos.

@tudor-da all that said, I am ok if you just do the template method, but happy to hear your thoughts on that matter!
cc @rautenrieth-da

Reply:
Thanks for the suggestion, @nmarton-da . I'll amend with your suggestion. I agree that `pruneEvents` needs a breakdown. Left 2 TODOs.